### ISSUE TITLE
Polar Flow Integration

### ISSUE THREAD

**Issue by liversedge** [GoldenCheetah/GoldenCheetah#2396]
Add support for Polar Flow integration (download only at this point).

**Comment by liversedge** [comment]
**Template code**

Add the implementation to the codebase, using TodaysPlan as a base to develop and enhance.

cbe4012000c49c688f782b686ef74d0c607f4fb6

**Comment by liversedge** [comment]
API doesn't really fit the cloud service model used by other services.
Pushing into the parking lot until the API changes.

**Comment by TheSmartGerman** [comment]
You know this link: https://github.com/polarofficial/accesslink-example-python ?

**Comment by adriansuter** [comment]
Any updates on this?

**Comment by ssj2son-gohan-ne** [comment]
Hey :) 

Did some work. It's some 

[](https://github.com/ssj2son-gohan-ne/GoldenCheetah/commit/71dd9662580bf8cd07899ee22d2cb5e2240083dc)

What now works:
- https://www.polar.com/accesslink-api/?shell#training-data
- getting all samples from exercise(, if provided): https://www.polar.com/accesslink-api/?shell#exercise-sample-types
- "time" is not provided by Polar, so it is created by given data from heartrate-sample, because heartrate-sample is the the one sample always provided

What needs to be done:
- fixLapSwim, i am not able to swim at the moment
- fixSmartRecording, this needs to be checked if it works for polar flow samples the right way
- overall testing, verify downloaded and sampled data
- checking conversation factors: "PolarFlow::addSamples"

What does Polar not provide as sample:
- latitude / longitude -> download gpx-file and combine with samples

What would be nice for future
- download gpx-file: https://www.polar.com/accesslink-api/?shell#get-gpx
- using webhooks: https://www.polar.com/accesslink-api/?shell#webhooks
- download tcx-file: https://www.polar.com/accesslink-api/?shell#get-tcx
- download fit-file: https://www.polar.com/accesslink-api/?shell#get-fit-beta

**Comment by liversedge** [comment]
awesome stuff. when you have something to test raise a pull request.
is there an upload api too?

**Comment by ssj2son-gohan-ne** [comment]
I don't think so. (At least so i understood)
Only download for the moment.
Can not find any instructions for uploading..

https://www.polar.com/accesslink-api/?shell#transaction-model

Adding manual Activities is possible by using the webpage (flow.polar.com) or the PolarFlowApp.

Cycling Example:
-----------------------
`
{
	"RIDE":{
		"STARTTIME":"2021\/10\/24 11:06:39 UTC ",
		"RECINTSECS":1,
		"DEVICETYPE":"Polar Vantage V2 ",
		"IDENTIFIER":" ",
		"OVERRIDES":[
			{ "Average Heart Rate":{ "value":"148" }},
			{ "Maximum Heart Rate":{ "value":"189" }},
			{ "total_distance":{ "value":"40.072" }},
			{ "workout_time":{ "value":"6957" }}
		],
		"TAGS":{
			"Athlete":"Test New Athlete ",
			"Data":"TDS-HC-A-L-E--- ",
			"Detailed Sport Info":"CYCLING ",
			"Device":"Polar Vantage V2 ",
			"File Format":" ",
			"Filename":"2021_10_24_13_06_39.json ",
			"Month":"Oktober ",
			"Polar Exercise ID":"123456789 ",
			"Sport":"Bike ",
			"Start Time Utc Offset":"120 ",
			"SubSport":"CYCLING ",
			"Weekday":"So. ",
			"Workout Code":"Training ",
			"Year":"2021 ",
			"calories":"1476 ",
			"carbohydrate-percentage":"60 ",
			"fat-percentage":"37 ",
			"has route":"1 ",
			"protein-percentage":"3 "
		},
		"SAMPLES":[
			{ "SECS":0, "KM":0, "CAD":0, "KPH":10.1584, "HR":117, "ALT":301.9079895, "SLOPE":0, "TEMP":24.6 },
			{ "SECS":1, "KM":0, "CAD":0, "KPH":10.1584, "HR":117, "ALT":301.9079895, "SLOPE":0, "TEMP":24.6 },
			{ "SECS":2, "KM":0.0021, "CAD":0, "KPH":9.9991, "HR":116, "ALT":301.9079895, "SLOPE":0, "TEMP":24.6 },
			{ "SECS":3, "KM":0.0042, "CAD":0, "KPH":9.7443, "HR":117, "ALT":302.82199097, "SLOPE":0, "TEMP":24.6 },
...
			{ "SECS":6950, "KM":40.0572, "CAD":0, "KPH":9.1389, "HR":142, "ALT":315.9289856, "SLOPE":-1.57586, "TEMP":22.2 },
			{ "SECS":6951, "KM":40.0589, "CAD":0, "KPH":9.1389, "HR":142, "ALT":315.9289856, "SLOPE":-1.57586, "TEMP":22.2 },
			{ "SECS":6952, "KM":40.0614, "CAD":0, "KPH":6.224, "HR":141, "ALT":315.9289856, "SLOPE":-1.57586, "TEMP":22.2 }
		]
	}
}
`

Running Example:
-----------------------
`
{
	"RIDE":{
		"STARTTIME":"2021\/10\/20 14:09:23 UTC ",
		"RECINTSECS":1,
		"DEVICETYPE":"Polar Vantage V ",
		"IDENTIFIER":" ",
		"OVERRIDES":[
			{ "Average Heart Rate":{ "value":"158" }},
			{ "Maximum Heart Rate":{ "value":"173" }},
			{ "total_distance":{ "value":"6.434" }},
			{ "workout_time":{ "value":"2370" }}
		],
		"TAGS":{
			"Athlete":"Test New Athlete ",
			"Data":"TDSPHC-A-L-E--- ",
			"Detailed Sport Info":"RUNNING ",
			"Device":"Polar Vantage V ",
			"File Format":" ",
			"Filename":"2021_10_20_16_09_23.json ",
			"Gear":"Shoe ",
			"Month":"Oktober ",
			"Polar Exercise ID":"123456789 ",
			"Sport":"Run ",
			"Start Time Utc Offset":"120 ",
			"SubSport":"RUNNING ",
			"Weekday":"Mi. ",
			"Workout Code":"Training ",
			"Year":"2021 ",
			"calories":"578 ",
			"has route":"1 ",
			"training load":"112 "
		},
		"SAMPLES":[
			{ "SECS":0, "KM":0, "WATTS":106, "CAD":0, "KPH":4.37, "HR":109, "ALT":307.85198975, "SLOPE":0, "TEMP":23 },
			{ "SECS":1, "KM":0, "WATTS":106, "CAD":0, "KPH":4.37, "HR":108, "ALT":307.85198975, "SLOPE":0, "TEMP":23 },
			{ "SECS":2, "KM":0.0019, "WATTS":106, "CAD":0, "KPH":4.29, "HR":108, "ALT":307.85198975, "SLOPE":0, "TEMP":23 },
...
			{ "SECS":2368, "KM":6.4264, "WATTS":336, "CAD":82, "KPH":12.62, "HR":159, "ALT":308.76599121, "SLOPE":2.40527, "TEMP":16.8 },
			{ "SECS":2369, "KM":6.4292, "WATTS":326, "CAD":81, "KPH":12.24, "HR":158, "ALT":308.76599121, "SLOPE":2.40527, "TEMP":16.8 },
			{ "SECS":2370, "KM":6.4312, "WATTS":316, "CAD":78, "KPH":11.84, "HR":158, "ALT":308.76599121, "SLOPE":2.40527, "TEMP":16.8 }
		]
	}
}

`
GPX need to be inserted
You can download a GPX from Polar, but yet i do not what to code therefore in GC...
Can you give me an advice?

**Comment by amtriathlon** [comment]
> GPX need to be inserted
> You can download a GPX from Polar, but yet i do not what to code therefore in GC...
> Can you give me an advice?

To download a file see Today’s Plan or Dropbox examples.
You could do that first to have a RideFile with GPS data, instead of creating an empty one, then addSamples should use appendOrUpdatePoint  instead of appendPoint to add remaining data.

**Comment by ssj2son-gohan-ne** [comment]
Is it possible to paste QByteArray directly to create a Ridefile?
so that there is no need to create a gpx-file.

Response from Polar Flow: https://www.polar.com/accesslink-api/?shell#get-gpx

<code>
**QByteArray
PolarFlow::getGPX(RideFile* ret, QString remoteid)
{
    // https://www.polar.com/accesslink-api/?shell#get-gpx
    qDebug() << "get GPX";

    printd("PolarFlow::addSamples\n");

    qDebug() << "get GPX RemoteID:" << remoteid;

    QString token = getSetting(GC_POLARFLOW_TOKEN, "").toString();

    QString transaction_id_Str = getSetting(GC_POLARFLOW_TRANSACTION_ID, "").toString();
    qDebug() << "get GPX transaction_id_String:" << transaction_id_Str;

    QString exercise_id_Str = getSetting(GC_POLARFLOW_EXERCISE_ID, "").toString();
    qDebug() << "get GPX exercise_id_String: " << exercise_id_Str;

 // RequestURL: GET https://www.polaraccesslink.com/v3/users/{user-id}/exercise-transactions/{transaction-id}/exercises/{exercise-id}/gpx
    QString get_GPX_Str;
            get_GPX_Str = QString("%1/v3/users/%2/exercise-transactions/%3/exercises/%4/gpx")
               .arg(getSetting(GC_POLARFLOW_URL, "https://www.polaraccesslink.com").toString())
               .arg(getSetting(GC_POLARFLOW_USER_ID, "").toString())
               .arg(getSetting(GC_POLARFLOW_TRANSACTION_ID, "").toString())
               .arg(remoteid);

    qDebug() << "get GPX - URL-List Samples:" << get_GPX_Str.toStdString().c_str();

    // request using the bearer token, application/gpx+xml
    QNetworkRequest request(get_GPX_Str);
                    request.setRawHeader("Authorization", (QString("Bearer %1").arg(token)).toLatin1());
                    request.setRawHeader("Accept", "application/gpx+xml");

    // make request
    QNetworkReply *reply = nam->get(request);
    printd("response : %s\n", get_GPX_Str.toStdString().c_str());
 // Example: Reply List of URLs /v3/users/{user-id}/exercise-transactions/{transaction-id}/exercises/{exercise-id}/gpx

    // blocking request
    QEventLoop loop;
    connect(reply, SIGNAL(finished()), &loop, SLOT(quit()));
    loop.exec();

    // if successful, lets unpack
    int statusCode = reply->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
    printd("fetch response: %d: %s\n", reply->error(), reply->errorString().toStdString().c_str());
    qDebug() << "get GPX - Samples - HTTP response status code: " << statusCode;

    QByteArray get_GPX_Ary = reply->readAll();

    printd("page : %s\n", get_GPX_Ary.toStdString().c_str());
    qDebug() << "get GPX - QByteArray:" << get_GPX_Ary.toStdString().c_str();

    qDebug() << "get GPX - close";

    return get_GPX_Ary;
}**
</code>

GPX is shortend
<code>
**Debug: get GPX - URL-List Samples: https://www.polaraccesslink.com/v3/users/1/exercise-transactions/123456789/exercises/987654321/gpx
Debug: get GPX - Samples - HTTP response status code:  200

Debug: get GPX - QByteArray: <?xml version="1.0" encoding="UTF-8"?><gpx xmlns="http://www.topografix.com/GPX/1/1" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="1.1" creator="Polar Accesslink" xsi:schemaLocation="http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd"><metadata><author><name>Polar Accesslink</name></author><time>2021-11-09T17:29:05.587Z</time><bounds minlat="1.1" minlon="1.1" maxlat="1.2" maxlon="1.2"></bounds></metadata><trk><trkseg><trkpt lat="1.1" lon="1.1"><ele>280.0</ele><time>2021-11-09T12:30:21.465Z</time></trkpt>
....
<trkpt lat="1.3" lon="1.3"><ele>312.0</ele><time>2021-11-09T13:59:00.465Z</time></trkpt></trkseg></trk></gpx>

Debug: get GPX - close**
</code>

**Comment by amtriathlon** [comment]
> Is it possible to paste QByteArray directly to create a Ridefile?

Not that I know

> so that there is no need to create a gpx-file.

Is there any advantage of downloading a gpx file and adding sample data, instead of just downloading a fit (or tcx) file?

**Comment by ssj2son-gohan-ne** [comment]
Good question.
Syncing changes instead of making the twice?!?!

I hope Polar will provide more information in there api as given in standardized file formats, for example at the moment training load, calories used energy, carbohydrate-percentage, protein-percentage and fat-percentage...

On Device you can set bike one, two or three. Taking this information for your gear.

**Comment by amtriathlon** [comment]
IMO, if the samples API doesn't provide additional information to the one included in the FIT (or TCX) files it would be far easier just to download one of them, similar to what Today's Plan interface does with JSON files.

OTOH, if additional metrics not derivable from the files are available via the API they can be added as metadata fields, we already are doing this with Garmin specific fields when importing FIT files.

**Comment by ssj2son-gohan-ne** [comment]
so only mixing the information and taking file-based information as basic / starting point ?!?
i did not think in this way ...

**Comment by amtriathlon** [comment]
What do you want to do? I mean with the information actually provided by the Polar API, not ideally.

**Comment by amtriathlon** [comment]
I am not sure I understand what you want to do.

What I am proposing is similar to the Today's Plan case, where the downloaded file (json in this case but in can be any of the supported formats) is augmented with metadata fields:

https://github.com/GoldenCheetah/GoldenCheetah/blob/cabe078453cbe8e136b8adf8d4187a5be878671b/src/Cloud/TodaysPlan.cpp#L569

**Comment by ssj2son-gohan-ne** [comment]
Thanks, Ale (:

I hope swimming is better imported by syncing directly from Polar Flow.
Importing TCX-SwimFile needs many afterwork.

And integration of Polar Flow woke my interest :)

**Comment by amtriathlon** [comment]
> I hope swimming is better imported by syncing directly from Polar Flow.
> Importing TCX-SwimFile needs many afterwork.

For pool swims FIT format is far better than TCX since it can include length records, but I don’t know if Polar is generating them, or providing length by length information otherwise in their API.

**Comment by ssj2son-gohan-ne** [comment]
fit files are provided by polar in two ways
https://www.polar.com/accesslink-api/?shell#get-fit-beta
https://www.polar.com/accesslink-api/?shell#get-exercise-fit

i don't know the differencies in fit-file, yet

the difference between the two gathering methods is via transaction-id and without

need to try

**Comment by ssj2son-gohan-ne** [comment]
Hey :)

is it possible to autoimport the downloaded tcx-file ?

RideImportWizard *rideImportFile = new RideImportWizard(fileName_Lst, context);
                               rideImportFile->process();
                               rideImportFile->isAutoImport();

[](https://github.com/ssj2son-gohan-ne/GoldenCheetah/commit/b81386dc2591bb22601ab448fd42c60d9f3cba67#diff-a5e3571c9c2a895f2d9454e078db2e4a2ef2193b4838a271d9d49243b1dd2784R1531
)

Is it possible to merge files on shared series ?

https://github.com/ssj2son-gohan-ne/GoldenCheetah/commit/b81386dc2591bb22601ab448fd42c60d9f3cba67#diff-a5e3571c9c2a895f2d9454e078db2e4a2ef2193b4838a271d9d49243b1dd2784R1740

**Comment by amtriathlon** [comment]
There’s no need to use Rideimportwizard, the cloud services API is intended to manage the process, see https://github.com/GoldenCheetah/GoldenCheetah/wiki/Using-the-Cloud-Services-API
As commented previously, Today’s Plan downloader is an example of downloading a file, amend it with additional data, and return the result.

**Comment by ssj2son-gohan-ne** [comment]
Thanks for the advise ( a second time) (:

Okay got it to go , now it first loads the tcx file and creates a RideFile.
Now i'm stucking at "appendOrUpdatePoint"

https://github.com/ssj2son-gohan-ne/GoldenCheetah/blob/Polar-Flow-Import-Only/src/Cloud/PolarFlow.cpp#L1040

simply changing from "appendPoint" to "appendOrUpdatePoint" does so easily as expected...

**Comment by ssj2son-gohan-ne** [comment]
Now the first working version for appendOrUpdatePoint
some work has to be done, structure works so far.

https://github.com/ssj2son-gohan-ne/GoldenCheetah/commit/83ff7f88f7f4a44fe93b00546967d2075786c2ef

**Comment by ssj2son-gohan-ne** [comment]
Hey, happy new year :)

Lately i decided to follow your advice @amtriathlon and do only the import of acquired tcx-file and add only additional information's available.
appendOrUpdate Points with Samples is parked for now, because my training starts to get more intensive and there is no more time hours of coding.

Here is my last commit:
https://github.com/ssj2son-gohan-ne/GoldenCheetah/commit/6a321d00c6e1d2c56e415248fa05d704b7d5959e

maybe so on could look over it and afterwards i will start a pull request...

**Comment by amtriathlon** [comment]
Hi, great!

WRT the general architecture, likely it is not a good idea to keep the transaction open, I think you should commit it in the [close method](https://github.com/GoldenCheetah/GoldenCheetah/wiki/Using-the-Cloud-Services-API#void-close), which is executed after all files selected for sync are downloaded.

I have not reviewed the code in detail since it is hard to follow with so many back-and-forth in several commits, I think you should do a cleanup removing all the commented out code, avoiding the logging duplication (`#define POLARFLOW_DEBUG true` enables printd, threre is no need to add redundant qDebug) and then git rebase and/or make a Pull Request so it can be reviewed at once.

Thank you!

**Comment by ssj2son-gohan-ne** [comment]
Hey :)

changed qDebug() lines to printd() for ones i know the syntax for
commented all qDebug() lines out for the ones i do not know the syntax for - TODO List

https://github.com/ssj2son-gohan-ne/GoldenCheetah/tree/Polar-Flow-Import-Only

Will now make a Pull Request, so that you can review the code in whole

:)

**Comment by Alexerson** [comment]
Hello, I just discovered GoldenCheetah and uses my Polar watch for swimming (alongside biking and running) and discovered that importing swimming data is painful (mostly because the TCX generated by Polar Flow is garbage).

Completing the import directly from the API would be great! Is there any progress being made lately? I can see some movements here: https://github.com/ssj2son-gohan-ne/GoldenCheetah/commits/Polar-Flow-Exercises-30Days
Anything I can help with (I can provide data, test, or code, depending on the needs).
